File: realm-transformer/src/main/java/io/realm/transformer/UrlEncodedAnalytics.java
package io.realm.transformer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.SocketException;
import java.net.URL;
import java.security.NoSuchAlgorithmException;



public class UrlEncodedAnalytics {

    private String prefix;
    private String suffix;

    public UrlEncodedAnalytics(String prefix, String suffix) {
        this.prefix = prefix;
        this.suffix = suffix;
    }

    public void execute(RealmAnalytics analytics) {
        try {
            URL url = getUrl(analytics);

            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.connect();
            connection.getResponseCode();
        } catch (Exception ignored) {
        }
    }

    private URL getUrl(RealmAnalytics analytics) throws
            MalformedURLException,
            SocketException,
            NoSuchAlgorithmException,
            UnsupportedEncodingException {
        return new URL(prefix + Utils.base64Encode(analytics.generateJson()) + suffix);
    }

    public static class MixPanel extends UrlEncodedAnalytics {
        private static final String ADDRESS_PREFIX = "https://api.mixpanel.com/track/?data=";
        private static final String ADDRESS_SUFFIX = "&ip=1";

        public MixPanel() {
            super(ADDRESS_PREFIX, ADDRESS_SUFFIX);
        }
    }

    public static class Segment extends UrlEncodedAnalytics {
        private static final String ADDRESS_PREFIX =
                "https://webhooks.mongodb-realm.com/api/client/v2.0/app/realmsdkmetrics-zmhtm/service/metric_webhook/incoming_webhook/metric?data=";
        private static final String ADDRESS_SUFFIX = "";

        public Segment() {
            super(ADDRESS_PREFIX, ADDRESS_SUFFIX);
        }
    }

}

File: realm-transformer/src/main/kotlin/io/realm/analytics/UrlEncodedAnalytics.kt
/*
 * Copyright 2021 Realm Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.realm.analytics

import io.realm.transformer.Utils
import java.io.UnsupportedEncodingException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.SocketException
import java.net.URL
import java.security.NoSuchAlgorithmException

class UrlEncodedAnalytics private constructor(private val prefix: String, private val suffix: String) {

    /**
     * Send the analytics event to the server.
     */
    fun execute(analytics: AnalyticsData) {
        try {
            val url = getUrl(analytics)
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connect()
            connection.responseCode
        } catch (ignored: Exception) {
        }
    }

    @Throws(
        MalformedURLException::class,
        SocketException::class,
        NoSuchAlgorithmException::class,
        UnsupportedEncodingException::class
    )
    private fun getUrl(analytics: AnalyticsData): URL {
        logger.error("Sending: \n${analytics.generateJson()}")
        return URL(prefix + Utils.base64Encode(analytics.generateJson()) + suffix)
    }

    companion object {
        fun create(): UrlEncodedAnalytics {
            val ADDRESS_PREFIX =
                "https://webhooks.mongodb-realm.com/api/client/v2.0/app/realmsdkmetrics-zmhtm/service/metric_webhook/incoming_webhook/metric?data="
            val ADDRESS_SUFFIX = ""
            return UrlEncodedAnalytics(ADDRESS_PREFIX, ADDRESS_SUFFIX)
        }
    }
}